rfcs/20200921-pickle-for-keras.md

```diff
@@ -0,0 +1,311 @@
+[tensorflow#39609].
+  
+[dask-ml#534]:https://github.com/dask/dask-ml/issues/534
+[SO#51110834]:https://stackoverflow.com/questions/51110834/cannot-pickle-dill-a-keras-object
+[SO#54070845]:https://stackoverflow.com/questions/54070845/how-to-pickle-keras-custom-layer
+[SO#59872509]:https://stackoverflow.com/questions/59872509/how-to-export-a-model-created-from-kerasclassifier-and-gridsearchcv-using-joblib
+[SO#37984304]:https://stackoverflow.com/questions/37984304/how-to-save-a-scikit-learn-pipline-with-keras-regressor-inside-to-disk
+[SO#48295661]:https://stackoverflow.com/questions/48295661/how-to-pickle-keras-model
+[skper]:https://scikit-learn.org/stable/modules/model_persistence.html#persistence-example
+[TF#33204]:https://github.com/tensorflow/tensorflow/issues/33204
+[TF#34697]:https://github.com/tensorflow/tensorflow/issues/34697
+
+[tensorflow#39609]:https://github.com/tensorflow/tensorflow/pull/39609
+[SciKeras]:https://github.com/adriangb/scikeras
+[skp]:https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/utils/estimator_checks.py#L1523-L1524
+
+<!--
+StackOverflow questions where `Model.save` would not work:
+
+* [SO#40396042](https://stackoverflow.com/questions/40396042/how-to-save-scikit-learn-keras-model-into-a-persistence-file-pickle-hd5-json-ya)
+  
+Examples that could be resolved using `Model.save` (but the user tried pickle first):
+
+* [SO #51878627](https://stackoverflow.com/questions/51878627/pickle-keras-ann)
+-->
+
+## User Benefit
+
+> How will users (or other contributors) benefit from this work? What would be the headline in the release notes or blog post?
+
+One blog post headline: "Keras models can be used with the advanced
+hyperparameter optimization techniques found in Dask-ML and Ray Tune." This has
+already been mentioned in "Framework support" of [a Dask blog post][dbp]
+comparing Dask-ML's hyperparameter optimization with Ray's tune-sklearn.
+
+[dbp]:https://blog.dask.org/2020/08/06/ray-tune#framework-support
+
+Users will also benefit with easier usage; they won't run into any of these
+errors:
+
+* People try to save Scikit-Learn meta-estimators with Keras components using
+  the serialization libraries Joblib or Dill. 
+  This fails because Keras models can not be serialized without a custom
+  method. Examples include [SO#59872509], [SO#37984304] and
+  [SO#48295661], and [SO#51110834].
+* Using custom parallelization strategies requires serialization support through
+  Pickle; however, many parallelization libraries don't
+  special case Keras models (e.g, Joblib). Relevant errors are most common in hyperparameter
+  optimization with Scikit-Learn's parallelization through Joblib
+  ([TF#33204] and [TF#34697]) or parallelization through Dask ([dask-ml#534]).
+* Lack of Pickle support can complicate saving training history like in
+  (the poorly asked) [SO#54070845].
+
+This RFC would resolve these issues.
+
+## Design Proposal
```

Comment:
What is the list of base classes that you expect will need to implement the `__reduce__` method?

Reply:
I think a good starting point would be:
- `tf.keras.Model`
- `tf.keras.metrics.Metric`
- `tf.keras.losses.Loss`
- `tf.keras.optimizers.Optimizer`

I think layers are usually serialized as part of a `Model`; I don't think there's much of a use case for implementing `__reduce__` for individual layers.

But generally, all classes that already have a `serialize` and `deserialize` method could benefit getting a `__reduce__` that just references those, but there is no reason why they have all be implemented together (or at all).